src/GraphEdgeSolver.cpp

```diff
@@ -20,30 +20,63 @@
 
 #include "GraphEdgeSolver.hpp"
 
+#include "LocalEigenSolver.hpp"
+
 namespace smoothg
 {
 
 GraphEdgeSolver::GraphEdgeSolver(const SparseMatrix& M, const SparseMatrix& D)
-    : GraphEdgeSolver(M.GetData(), D)
+    : is_diag_(IsDiag(M))
 {
-}
+    assert(M.Rows() > 0);
+    assert(D.Rows() > 0);
 
-GraphEdgeSolver::GraphEdgeSolver(const std::vector<double>& M_data,
-                                 const SparseMatrix& D)
-    : MinvDT_(D.Transpose()), vect_sol_(D.Rows(), 0.0)
-{
-    MinvDT_.InverseScaleRows(M_data);
+    if (is_diag_)
+    {
+        rhs_ = BlockVector({0, 0, D.Rows()});
+        sol_ = BlockVector({0, 0, D.Rows()});
+
+        MinvDT_ = D.Transpose();
+        MinvDT_.InverseScaleRows(M.GetData());
+
+        SparseMatrix A = D.Mult(MinvDT_);
+        A.EliminateRowCol(0);
+
+        block_Ainv_ = SparseSolver(std::move(A));
+    }
+    else
+    {
+        rhs_ = BlockVector({0, M.Rows(), M.Rows() + D.Rows()});
+        sol_ = BlockVector({0, M.Rows(), M.Rows() + D.Rows()});
+
+        linalgcpp::BlockMatrix<double> block({0, M.Rows(), M.Rows() + D.Rows()});
+
+        int elim_dof = 0;
+
+        CooMatrix W_coo(D.Rows(), D.Rows());
+        W_coo.Add(elim_dof, elim_dof, 1);
+
+        SparseMatrix W = W_coo.ToSparse();
+
+        SparseMatrix D_elim(D);
+        D_elim.EliminateRow(elim_dof);
+        SparseMatrix DT_elim = D_elim.Transpose();
+
+        block.SetBlock(0, 0, M);
+        block.SetBlock(0, 1, DT_elim);
+        block.SetBlock(1, 0, D_elim);
+        block.SetBlock(1, 1, W);
```

Comment:
An easier way would be to set blocks `M`, `DT`, and `D` in `block`, and eliminate `elim_dof = M.Rows()` of the combined matrix.

Reply:
Actually, should there be some assert or warning print in `SparseMatrix<T>::EliminateRowCol` when diagonal entry (at least the one being eliminated) is not in the sparsity pattern? This will add complexity to the function though. If not, maybe state this assumption in the documentation of the function. 

Similar issue also for `SparseMatrix<T>::AddDiag`.

Reply:
This ML code is in no way complete or optimized. Moving into the block matrix will avoid all the copies.
Equivalent idea of just moving everything: 
```c++
        block.SetBlock(0, 0, std::move(M));
        block.SetBlock(0, 1, std::move(DT_elim));
        block.SetBlock(1, 0, std::move(D_elim));
        block.SetBlock(1, 1, std::move(W));
```